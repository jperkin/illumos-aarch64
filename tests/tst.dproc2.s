/*
 * This file and its contents are supplied under the terms of the
 * Common Development and Distribution License ("CDDL"), version 1.0.
 * You may only use this file in accordance with the terms of version
 * 1.0 of the CDDL.
 *
 * A full copy of the text of the CDDL should have accompanied this
 * source.  A copy of the CDDL is also available via the Internet at
 * http://www.illumos.org/license/CDDL.
 */

/*
 * Copyright 2017 Joyent, Inc.
 */

/*
 * C3.5.8 Data-processing (2 source)
 */
.macro dproc op
	\op	w0,  w15, w30
	\op	x0,  x15, x30
	\op	wzr, wzr, wzr
	\op	xzr, xzr, xzr
.endm
.macro dproc_crc op xreg=0
.if \xreg
	\op	w0,  w15, x30
	\op	wzr, wzr, xzr
.else
	\op	w0,  w15, w30
	\op	wzr, wzr, wzr
.endif
.endm

.text
.arch armv8-a+crc
.globl libdis_test
.type libdis_test, @function
libdis_test:
	/*
	 * C5.6.214 UDIV
	 */
	dproc udiv
	/*
	 * C5.6.160 SDIV
	 */
	dproc sdiv
	/*
	 * C5.6.115 LSLV
	 * C5.6.113 LSL (register) (alias)
	 */
	dproc lslv
	/*
	 * C5.6.118 LSRV
	 * C5.6.116 LSR (register) (alias)
	 */
	dproc lsrv
	/*
	 * C5.6.17 ASRV
	 * C5.6.15 ASR (register) (alias)
	 */
	dproc asrv
	/*
	 * C5.6.154 RORV
	 * C5.6.153 ROR (register) (alias)
	 */
	dproc rorv
	/*
	 * C5.6.48 CRC32B, CRC32H, CRC32W, CRC32X
	 */
	dproc_crc crc32b
	dproc_crc crc32h
	dproc_crc crc32w
	dproc_crc crc32x 1
	/*
	 * C5.6.49 CRC32CB, CRC32CH, CRC32CW, CRC32CX
	 */
	dproc_crc crc32cb
	dproc_crc crc32ch
	dproc_crc crc32cw
	dproc_crc crc32cx 1
.size libdis_test, [.-libdis_test]
