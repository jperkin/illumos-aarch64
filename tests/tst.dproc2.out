    libdis_test:      e0 09 de 1a  udiv w0, w15, w30
    libdis_test+0x4:  e0 09 de 9a  udiv x0, x15, x30
    libdis_test+0x8:  ff 0b df 1a  udiv wzr, wzr, wzr
    libdis_test+0xc:  ff 0b df 9a  udiv xzr, xzr, xzr
    libdis_test+0x10: e0 0d de 1a  sdiv w0, w15, w30
    libdis_test+0x14: e0 0d de 9a  sdiv x0, x15, x30
    libdis_test+0x18: ff 0f df 1a  sdiv wzr, wzr, wzr
    libdis_test+0x1c: ff 0f df 9a  sdiv xzr, xzr, xzr
    libdis_test+0x20: e0 21 de 1a  lsl w0, w15, w30
    libdis_test+0x24: e0 21 de 9a  lsl x0, x15, x30
    libdis_test+0x28: ff 23 df 1a  lsl wzr, wzr, wzr
    libdis_test+0x2c: ff 23 df 9a  lsl xzr, xzr, xzr
    libdis_test+0x30: e0 25 de 1a  lsr w0, w15, w30
    libdis_test+0x34: e0 25 de 9a  lsr x0, x15, x30
    libdis_test+0x38: ff 27 df 1a  lsr wzr, wzr, wzr
    libdis_test+0x3c: ff 27 df 9a  lsr xzr, xzr, xzr
    libdis_test+0x40: e0 29 de 1a  asr w0, w15, w30
    libdis_test+0x44: e0 29 de 9a  asr x0, x15, x30
    libdis_test+0x48: ff 2b df 1a  asr wzr, wzr, wzr
    libdis_test+0x4c: ff 2b df 9a  asr xzr, xzr, xzr
    libdis_test+0x50: e0 2d de 1a  ror w0, w15, w30
    libdis_test+0x54: e0 2d de 9a  ror x0, x15, x30
    libdis_test+0x58: ff 2f df 1a  ror wzr, wzr, wzr
    libdis_test+0x5c: ff 2f df 9a  ror xzr, xzr, xzr
    libdis_test+0x60: e0 41 de 1a  crc32b w0, w15, w30
    libdis_test+0x64: ff 43 df 1a  crc32b wzr, wzr, wzr
    libdis_test+0x68: e0 45 de 1a  crc32h w0, w15, w30
    libdis_test+0x6c: ff 47 df 1a  crc32h wzr, wzr, wzr
    libdis_test+0x70: e0 49 de 1a  crc32w w0, w15, w30
    libdis_test+0x74: ff 4b df 1a  crc32w wzr, wzr, wzr
    libdis_test+0x78: e0 4d de 9a  crc32x w0, w15, x30
    libdis_test+0x7c: ff 4f df 9a  crc32x wzr, wzr, xzr
    libdis_test+0x80: e0 51 de 1a  crc32cb w0, w15, w30
    libdis_test+0x84: ff 53 df 1a  crc32cb wzr, wzr, wzr
    libdis_test+0x88: e0 55 de 1a  crc32ch w0, w15, w30
    libdis_test+0x8c: ff 57 df 1a  crc32ch wzr, wzr, wzr
    libdis_test+0x90: e0 59 de 1a  crc32cw w0, w15, w30
    libdis_test+0x94: ff 5b df 1a  crc32cw wzr, wzr, wzr
    libdis_test+0x98: e0 5d de 9a  crc32cx w0, w15, x30
    libdis_test+0x9c: ff 5f df 9a  crc32cx wzr, wzr, xzr
